src/ClientData/ScopeIdProvider.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ClientData/ScopeIdProvider.h"
+
+#include <absl/flags/flag.h>
+
+#include <algorithm>
+#include <cstdint>
+#include <memory>
+
+#include "ClientData/Constants.h"
+#include "ClientFlags/ClientFlags.h"
+#include "GrpcProtos/Constants.h"
+
+namespace orbit_client_data {
+
+[[nodiscard]] std::unique_ptr<NameEqualityScopeIdProvider> NameEqualityScopeIdProvider::Create(
+    const orbit_grpc_protos::CaptureOptions& capture_options) {
+  const auto& instrumented_functions = capture_options.instrumented_functions();
+  const uint64_t max_id =
+      instrumented_functions.empty()
+          ? 0
+          : std::max_element(
+                std::begin(instrumented_functions), std::end(instrumented_functions),
+                [](const auto& a, const auto& b) { return a.function_id() < b.function_id(); })
+                ->function_id();
+  return std::unique_ptr<NameEqualityScopeIdProvider>(new NameEqualityScopeIdProvider(max_id + 1));
+}
+
+NameEqualityScopeIdProvider::NameEqualityScopeIdProvider(uint64_t start_id) : next_id_(start_id) {}
```

Comment:
Nit/personal preference: I'd prefer constructors only using type initializers to be implemented directly in the header.